data_structures/src/chain/priority.rs

```diff
@@ -0,0 +1,916 @@
+        // Update highest
+        if priority > self.vtt_highest {
+            self.vtt_highest = priority;
+        }
+        // Update lowest
+        if let Some(vtt_lowest) = &self.vtt_lowest {
+            if &priority < vtt_lowest {
+                self.vtt_lowest = Some(priority);
+            }
+        } else if priority != 0 {
+            self.vtt_lowest = Some(priority);
+        }
+    }
+}
+
+impl fmt::Debug for Priorities {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "[DRT](↑{:<10}, ↓{:<10}) | [VTT](↑{:<10}, ↓{:<10})",
+            self.drt_highest,
+            self.drt_lowest.unwrap_or_default(),
+            self.vtt_highest,
+            self.vtt_lowest.unwrap_or_default()
+        )
+    }
+}
+
+/// A whole set of estimates for priority of DRT and VTT transactions.
+///
+/// Each estimate contains values for 5 different tiers of priority:
+/// - Stinky
+/// - low
+/// - Medium
+/// - High
+/// - Opulent
+#[derive(Clone, Debug, Default, Deserialize, Eq, PartialEq, Serialize)]
+pub struct PrioritiesEstimate {
+    pub drt_stinky: PriorityEstimate,
+    pub drt_low: PriorityEstimate,
+    pub drt_medium: PriorityEstimate,
+    pub drt_high: PriorityEstimate,
+    pub drt_opulent: PriorityEstimate,
+    pub vtt_stinky: PriorityEstimate,
+    pub vtt_low: PriorityEstimate,
+    pub vtt_medium: PriorityEstimate,
+    pub vtt_high: PriorityEstimate,
+    pub vtt_opulent: PriorityEstimate,
+}
+
+impl fmt::Display for PrioritiesEstimate {
+    #[allow(clippy::to_string_in_format_args)]
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            r#"╔══════════════════════════════════════════════════════════╗
+║ TRANSACTION PRIORITY ESTIMATION REPORT                   ║
+╠══════════════════════════════════════════════════════════╣
+║ Data request transactions                                ║
+╟──────────┬──────────────────┬────────────────────────────║
+║     Tier │ Priority         │ Time-to-block              ║
+╟──────────┼──────────────────┼────────────────────────────║
+║   Stinky │ {:<16} │ {:<25}  ║
+║      Low │ {:<16} │ {:<25}  ║
+║   Medium │ {:<16} │ {:<25}  ║
+║     High │ {:<16} │ {:<25}  ║
+║  Opulent │ {:<16} │ {:<25}  ║
+╠══════════════════════════════════════════════════════════╣
+║ Value transfer transactions                              ║
+╟──────────┬──────────────────┬────────────────────────────║
+║     Tier │ Priority         │ Time-to-block              ║
+╟──────────┼──────────────────┼────────────────────────────║
+║   Stinky │ {:<16} │ {:<25}  ║
+║      Low │ {:<16} │ {:<25}  ║
+║   Medium │ {:<16} │ {:<25}  ║
+║     High │ {:<16} │ {:<25}  ║
+║  Opulent │ {:<16} │ {:<25}  ║
+╚══════════════════════════════════════════════════════════╝"#,
+            // Believe it or not, these `to_string` are needed for proper formatting, hence the
+            // clippy allow directive above.
```

Comment:
So basically, this is the `Display` impl of the `Priority` type:

```
impl fmt::Display for Priority {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{:.3}", self.0)
    }
}
```

The problem is that it ignores all the flags passed to the formatter. So `{:<8}` is handled the same as `{}`. This is because `write!` calls `f.write_str`, and `f.write_str` is used to write raw strings, ignoring all the extra flags. So the write macro ignores the flags even if we manually convert the type to string:

```
// This is the same as write!(f, "{:.3}", self.0)
let three_decimals = format!("{:.3}", self.0);
write!(f, "{}", three_decimals)
```

There are ways to manually check the padding and alignment flags, for example using `f.align()`, but the easiest approach is to forward to the `Display` implementation of a related type, such as `String`:

```
let three_decimals = format!("{:.3}", self.0);
three_decimals.fmt(f)
```

But I haven't seen any uses of that code pattern, we always use `write!` or `f.write_str` to implement `Display`, and that ignores padding and alignment. So I would just leave it as it is and keep the allow clippy, because it makes sense to convert a type to a string and then change the padding of that string, so clippy is wrong in this case.